ppafm/ocl/field.py

```diff
@@ -37,89 +41,356 @@ def makeDivisibleUp( num, divisor ):
+                to the file path for each component, respectively.
+            clamp: float or None. If not None, all values greater than this are clamped to this value.
+        '''
+        file_head, ext = os.path.splitext(file_path)
+        if ext not in ['.xsf', '.npy']:
+            raise ValueError(f'Unsupported file extension `{ext}` for saving data grid.')
+        ext = ext[1:]
+        array = self.array.copy()
+        if len(self.shape) == 3:
+            if clamp: array[array > clamp] = clamp
+            io.save_scal_field(file_head, array.T, self.lvec, data_format=ext)
+        if len(self.shape) == 4:
+            assert self.shape[3] == 4, 'Wrong number of components'
+            if clamp:
+                io.limit_vec_field(array, Fmax=clamp)
+                array[:, :, :, 3][array[:, :, :, 3] > clamp] = clamp
+            array = array.transpose(2, 1, 0, 3)
+            io.save_vec_field(file_head, array[:, :, :, :3], self.lvec, data_format=ext)
+            io.save_scal_field(file_head+'_w', array[:, :, :, 3], self.lvec, data_format=ext)
+
+    def add_mult(self, array, scale=1.0, in_place=True, local_size=(32,), queue=None):
+        '''
+        Multiply the values of another data grid and add them to the values of this data grid.
+
+        Arguments:
+            array: DataGrid. Grid whose values to scale and add.
+            scale: float. Value by which values in array are multiplied.
+            in_place: bool. Whether to do operation in place or to create a new array.
+            local_size: tuple of a single int. Size of local work group on device.
+            queue: pyopencl.CommandQueue. OpenCL queue on which operation is performed.
+                Defaults to oclu.queue.
+
+        Returns:
+            grid_out: Same type as self. New data grid with result.
+        '''
+        array_in1 = self.cl_array
+        array_in2 = array.cl_array
+        queue = queue or oclu.queue
+        if in_place:
+            grid_out = self
+            array_out = array_in1
+            self._array = None # The current numpy array will be wrong after operation so reset it
+        else:
+            array_out = cl.Buffer(self.ctx, cl.mem_flags.READ_WRITE, size=array_in1.size)
+            array_type = type(self) # This way so inherited classes return their own class type
+            grid_out = array_type(array_out, lvec=self.lvec, shape=self.shape, ctx=self.ctx)
+        n = np.int32(array_in1.size / 4)
+        scale = np.float32(scale)
+        global_size = [int(np.ceil(n / local_size[0]) * local_size[0])]
+        cl_program.addMult(queue, global_size, local_size, array_in1, array_in2, array_out, n, scale)
+        return grid_out
+
+    def power_positive(self, p=1.2, in_place=True, local_size=(32,), queue=None):
+        '''
+        Raise every positive element in the grid into a power. Negative values are set to zero.
```

Comment:
`negative values are set to zero` seems a bit suspicious. I guess you assume grid contains electron density, which should be possitive everywhere, but due to numerical errors it is sometimes somewhere negative.

I see possible problem that if you set these negative values to zero, you change total amout of chrge in the system. I guess it is very small amout? But maybe it woul be better to renormalize it or something? I know that in some DFT codes using grids, there is renormalization for similar reason.

Reply:
Hmm, I was thinking that I just did it the same you did in `conv_rho.py`, but now I realize there is also a rescaling happening there at https://github.com/Probe-Particle/ppafm/blob/710d046686a2e9f0109556edc681db032f8f4131/ppafm/cli/conv_rho.py#L21-L24
This is more challenging to implement in a single kernel, since you would have to do two sums over the whole array. I think I will just open another issue for this.

Reply:
Added new issue #136.